### PR TITLE
Tdd lesson

### PR DESCRIPTION
**Pull request by CheezItMan**
## Description
I've updated the TDD lesson and setup a sample repo for starter code.  I'll have something for the teacher's edition soon.  I've changed it to spec-style testing, with a brief bit about asserts so they'll know what they are.  

## Todos
Are any of these items still in progress?
-  No

## Feedback Requested
Another set of eyes is welcome.

### CODE REVIEW

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 1]
```diff
@@ -60,24 +66,273 @@ true
...
 
-### Assertions
-An _assertion_ is the part of the _automated test suite_ that tells us whether our code did what we expected it to. Usually, _assertions_ are special methods that will evaluate whether an expression is `true`. If it is, the _assertion_ is true, the test passes (green test), otherwise it fails (red test).
 
-### Exercise: Write an `assert` method
-Let's build our own `assert` method in the `exponate_test.rb` file. Here's what it should look like:
+## Now Comes Minitest
+
+Writing your own tests with puts is wonderful, but it would be handy to have a standard way that developers can use to write tests on their code, a way that other developers understand.  The maintainers of the Ruby language have adopted a testing library called [Minitest](http://docs.seattlerb.org/minitest/) as the default standard for testing in Ruby & later Rails.  For the remainder of your time using Ruby at Ada, we will be using Minitest to write unit-tests for your code.  There are other testing frameworks, notably [RSpec](http://rspec.info/	), but the key concepts are similar.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: For the last sentence some thing like: "RSpec is another very common testing framework used along with Ruby and Rails. We won't be using it here at Ada but it's good to know about when you're browsing the internet for testing help."

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 2]
```diff
@@ -60,24 +66,273 @@ true
...
+### How To Use Minitest?
+
+Lets say we need to write a class which simulates a 6-sided die.  We could use such a module in any number of board games (Monopoly etc).  
+
+To start with we'll create a file to test the class.  We'll call it `die_test.rb`.  We will also require `minitest/autorun`
+
+```ruby
+# die_test.rb
+
+require 'minitest/autorun'
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: For the first time, will they need to gem install this?

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 3]
```diff
@@ -60,24 +66,273 @@ true
...
+
+
+### A Tale of Two Styles
+
+In the TDD World there are two styles of testing.  In the first more traditional method people use *assertions* which are statements that check if a value is what it should be.  The other method is a subset of TDD, called Behavior-Driven Development (BDD) which accomplishes the same thing in a more English-friendly fashion that even non-developers can understand. 
+
+
+#### Writing Your First Test - Assertion Style
+
+To start we'll need to create a TestCase class.  We create a set of unit tests by subclassing MiniTest::Unit::TestCase and add each set of tests in instance methods.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I think it'd be good to wrap the class names in the code tags here for readability

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 4]
```diff
@@ -60,24 +66,273 @@ true
...
+```
+# die_test.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+
+class TestDie < MiniTest::Unit::TestCase
+  def test_creation_of_die
+    @die = Die.new
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I think this should probably be a local variable instead of an instance variable at this point

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 5]
```diff
@@ -60,24 +66,273 @@ true
...
+
+We are now setup, but we haven't actually written a test yet.
+
+
+### A Tale of Two Styles
+
+In the TDD World there are two styles of testing.  In the first more traditional method people use *assertions* which are statements that check if a value is what it should be.  The other method is a subset of TDD, called Behavior-Driven Development (BDD) which accomplishes the same thing in a more English-friendly fashion that even non-developers can understand. 
+
+
+#### Writing Your First Test - Assertion Style
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: There are a bunch of steps in this section. Could we label each piece with a number or like a "First we ...", "second we..." header of some sort

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 6]
```diff
@@ -60,24 +66,273 @@ true
...
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+```
+
+Now we have our first **green/passing** test.  
+
+There are a [number of assertions](https://gist.github.com/rastasheep/4248006#Minitest::Unit::TestCase) in Minitest beyond the assert method.  
+
+#### Practice Exercise
+
+Now we need to test that we can roll the die.  Write a test that checks the roll method of the Die class.  When a die is rolled it should return a number between 1 and 6 inclusive.  You can use the `assert_operator` method.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: For this, they also need to write the method itself, right? The way it is worded right now it feels like it would get them to the "red" test but not necessarily writing the method in the Die class.

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 7]
```diff
@@ -60,24 +66,273 @@ true
...
+
+```ruby
+# die_spec.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+```
+
+Up till now this all looks the same, but next it changes a bit.  Instead of creating a class we'll write a `describe` block which identifies a group of tests that belong together, usually because they are testing the same thing.  Inside the `describe` block are `it` blocks.  Each `it` block is a test.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: until instead of till

Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I think another sentence or two explaining the "why" would be good before diving into each type of block.

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 8]
```diff
@@ -60,24 +66,273 @@ true
...
+  it "You can create an instance of Die" do
+    @die = Die.new
+
+    @die.class.must_equal Die
+  end
+end
+
+```
+
+Notice the line `@die.class.must_equal Die`.  Look at how english-friendly it is.  The idea with **BDD** is the that with our tests we are defining specifications or behaviors our code should follow.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: For the second sentence, maybe something more like "Observe how this reads more easily from left to right than our original assert test."

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 9]
```diff
@@ -60,24 +66,273 @@ true
...
+Check again with your partner when you are finished.  You can find a solution [here](https://github.com/AdaGold/die/blob/solution/specs/die_spec.rb)
+
+
+## Which Style Should I Use?
+
+So which style is better?  Functionally both do the same job, but the spec-style is more readable by non-technical people.  So we will, from this point on, use **Spec-style tests.**  However it is important for you to know that both exist and when you get to rails, a great deal more documentation exists for assert-style tests compared to spec-style.   
+
+## What Should I Test?
+
+More important that how you test your code is what you are testing.  If you're not testing the right things bugs can creep through your tests and into production code.  Many many many developers have trouble knowing what to test.   Here are some general guidelines.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: than instead of that

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 10]
```diff
@@ -60,24 +66,273 @@ true
...
+	*  If your method took an array as a parameter you would test it with an empty array, a one element array and a large array.  
+*  Think about how someone might misuse your method, check for invalid or weird input.  If someone can break your code... they will.
+
+## Organizing Code
+
+As our projects get larger putting all our code in the same folder gets messy, especially with test files.  So we will be setting up projects like this:
+
+![Project Structure](images/project_folder.png "project structure")
+
+We will be placing our code in the **lib** folder, and our test specs in the **specs** folder.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I feel like it might be more confusing to do this at the end once we have already gone through an example. What made you decide to do this rather than putting them in this file structure from the start?

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 11]
```diff
@@ -60,24 +66,273 @@ true
...
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+
+class TestDie < MiniTest::Unit::TestCase
+  def test_creation_of_die
+    @die = Die.new
+    # The class of @die should be Die
+    assert @die.class == Die, "There must be a Die class."
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I'm not sure where this observation should be presented, but we should talk about the fact that we leave parenthesis off of `assert` and `must` methods even though we don't ever lead parenthesis off of non-testing methods. (comes up in another example below and in the assertions table at the bottom_)

File: 00-programming-fundamentals/08-intro-to-automated-tests.md (hunk 10 above)
Comment (CheezItMan) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Mostly because Rake/organization is something that looked like we might want to skip and come back to at a later date.  Placing it at the end made it easier to skip if time, and too much material made it undesirable as an intro.

I can work it in earlier.

File: 00-programming-fundamentals/08-intro-to-automated-tests.md (hunk 2 above)
Comment (CheezItMan) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Done, I'd forgotten Minitest-reporters is not installed by default.

File: 00-programming-fundamentals/08-intro-to-automated-tests.md (hunk 5 above)
Comment (CheezItMan) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Done

File: 00-programming-fundamentals/08-intro-to-automated-tests.md (hunk 4 above)
Comment (CheezItMan) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Fixed

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 12]
```diff
@@ -60,24 +67,331 @@ true
...
+
+Spec style tests look very different, but read in a more English friendly fashion.  We'll create the test with a file named `die_spec.rb`.
+
+```ruby
+# die_spec.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+```
```
Comment (droberts-ada) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Do you need a `require 'minitest/spec'` here?

Comment (CheezItMan) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: @droberts-ada 

Not Anymore!

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 13]
```diff
@@ -60,24 +67,337 @@ true
...
+NameError: uninitialized constant TestDie::Die
+Did you mean?  Dir
+    die_test.rb:8:in `test_creation_of_die'
+
+1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
+```
+
+**This is a good thing.**  We have our first **red** test.  There's an error because we haven't created the Die class yet.
+
+Lets do so:
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I think this is maybe where step # 4 would be

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 14]
```diff
@@ -60,24 +67,337 @@ true
...
+Now we have our first **green/passing** test.  
+
+There are a [number of assertions](https://gist.github.com/rastasheep/4248006#Minitest::Unit::TestCase) in Minitest beyond the assert method.  
+
+
+#### A Word on Parentheses
+
+Check out the code above.  We are calling assert without using parentheses here:  `assert @die.class == Die, "There must be a Die class."`  Ruby doesn't FORCE you to put parentheses around a method's arguments but [the community-driven style guidelines](https://github.com/bbatsov/ruby-style-guide) suggest that it's good coding style to put parentheses around method arguments **except** for methods part of an internal Domain Specific Language (DSL), or basically the syntax of some kind of framework like... Minitest.  
+
+So you shouldn't put parentheses around the arguments to `assert` or later `must` method arguments, but you **should** around your own methods.
```
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Perfect! Thanks for adding this

File: 00-programming-fundamentals/08-intro-to-automated-tests.md (hunk 10 above)
Comment (kariabancroft) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Yeah I think I agree - i"m not sure we even need it in here for now? Especially since we will have special instructions for running weekend warrior. thoughts?

Comment (CheezItMan) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: I just felt a little awkward sending them out with weekend warrior without seeing that before in a class environment.

File: 00-programming-fundamentals/08-intro-to-automated-tests.md [hunk 15]
```diff
@@ -60,24 +67,337 @@ true
...
+*  Test your methods with edge case values.
+	*  For example if your method takes an integer as a parameter you would test it with a positive number, a negative number and zero.
+	*  If your method took an array as a parameter you would test it with an empty array, a one element array and a large array.  
+*  Think about how someone might misuse your method, check for invalid or weird input.  If someone can break your code... they will.
+
+## Organizing Code
+
+As our projects get larger putting all our code in the same folder gets messy, especially with test files.  So we will be setting up projects like this:
+
+![Project Structure](images/project_folder.png "project structure")
```
Comment (droberts-ada) [inline 00-programming-fundamentals/08-intro-to-automated-tests.md]: Not a big issue, but I've been trying to actively dissuade my students from using Finder. Could we show terminal output or something here?

image
